FernFlower-Patches/0023-Synthetic-getClass-Objects.requireNonNull-cleanup.patch

```diff
@@ -69,60 +72,98 @@ index 986a8a72ec138ea637d3446ed2e823f6100c31cc..767f6bdb92ecf0ba0bc4a437b8bf8482
-+        }
++      if ((!invocation.isStatic() &&
++           invocation.getName().equals("getClass") && invocation.getStringDescriptor().equals("()Ljava/lang/Class;")) // J8
++        || (invocation.isStatic() && invocation.getClassname().equals("java/util/Objects") && invocation.getName().equals("requireNonNull")
++            && invocation.getStringDescriptor().equals("(Ljava/lang/Object;)Ljava/lang/Object;"))) { // J9+
  
 -        List<Exprent> lstExprents = second.getAllExprents();
++        if (invocation.isSyntheticNullCheck()) {
++          return true;
++        }
++
 +        LinkedList<Exprent> lstExprents = new LinkedList<>();
          lstExprents.add(second);
  
 -        for (Exprent expr : lstExprents) {
-+        while (!lstExprents.isEmpty()){
++        final Exprent target;
++        if (invocation.isStatic()) { // Objects.requireNonNull(target) (J9+)
++          // detect target type
++          target = invocation.getLstParameters().get(0);
++        } else { // target.getClass() (J8)
++          target = invocation.getInstance();
++        }
++
++        while (!lstExprents.isEmpty()) {
 +          Exprent expr = lstExprents.removeFirst();
 +          lstExprents.addAll(expr.getAllExprents());
            if (expr.type == Exprent.EXPRENT_NEW) {
              NewExprent newExpr = (NewExprent)expr;
              if (newExpr.getConstructor() != null && !newExpr.getConstructor().getLstParameters().isEmpty() &&
 -                newExpr.getConstructor().getLstParameters().get(0).equals(invocation.getInstance())) {
-+              (newExpr.getConstructor().getLstParameters().get(0).equals(invocation.getInstance()) ||
-+                newExpr.getConstructor().getLstParameters().get(0).getExprType().equals(invocation.getInstance().getExprType()))) {
++              (newExpr.getConstructor().getLstParameters().get(0).equals(target) ||
++                isUnambiguouslySameParam(invocation.isStatic(), target, newExpr.getConstructor().getLstParameters()))) {
  
                String classname = newExpr.getNewType().value;
                ClassNode node = DecompilerContext.getClassProcessor().getMapRootClasses().get(classname);
+@@ -535,6 +571,27 @@ public class SimplifyExprentsHelper {
+     return false;
+   }
+ 
++  private static boolean isUnambiguouslySameParam(boolean isStatic, Exprent target, List<Exprent> parameters) {
++    boolean firstParamOfSameType = parameters.get(0).getExprType().equals(target.getExprType());
++    if (!isStatic) { // X.getClass()/J8, this is less likely to overlap with legitimate use
++      return firstParamOfSameType;
++    }
++    // Calling Objects.requireNonNull and discarding the result is a common pattern in normal code, so we have to be a bit more
++    // cautious about stripping calls when a constructor takes parameters of the instance type
++    // ex. given a class X, `Objects.requireNonNull(someInstanceOfX); new X(someInstanceOfX)` should not have the rNN stripped.
```

Comment:
Tho this is an okay assumption, looking at where the JDK actually emits the [null checks](https://github.com/AdoptOpenJDK/openjdk-jdk/blob/master/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java#L2910)
We have the following cases:

Lambdas: [Expr::method](https://github.com/AdoptOpenJDK/openjdk-jdk/blob/6eb91145823275248d049f8557462f09270c0b36/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LambdaToMethod.java#L502)
Lambdas: [LocalCall::new](https://github.com/AdoptOpenJDK/openjdk-jdk/blob/6eb91145823275248d049f8557462f09270c0b36/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LambdaToMethod.java#L936)
Inners: [new instance.Inner()](https://github.com/AdoptOpenJDK/openjdk-jdk/blob/6eb91145823275248d049f8557462f09270c0b36/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Lower.java#L2805) [2?](https://github.com/AdoptOpenJDK/openjdk-jdk/blob/6eb91145823275248d049f8557462f09270c0b36/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Lower.java#L3015)
[Switch on string with zero cases](https://github.com/AdoptOpenJDK/openjdk-jdk/blob/6eb91145823275248d049f8557462f09270c0b36/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Lower.java#L3706).

The [last one](https://github.com/AdoptOpenJDK/openjdk-jdk/blob/6eb91145823275248d049f8557462f09270c0b36/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransTypes.java#L681) is referenced in a few places. It seems to be adding a null check when creating new instances of classes inside other anon classes?

Anyways, for now this would be an acceptable 'get it working' patch. However the filter system should be more robust and address the actual areas in which the compiler outputs null checks.
Note: The old .getClass() check wasn't as restrictive as i'm saying. But this is more for reference and 'doing it right'